sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -198,16 +192,15 @@ object ConstantPropagation extends Rule[LogicalPlan] {
   private def safeToReplace(ar: AttributeReference, nullIsFalse: Boolean) =
     !ar.nullable || nullIsFalse
 
-  private def replaceConstants(condition: Expression, equalityPredicates: EqualityPredicates)
-    : Expression = {
-    val constantsMap = AttributeMap(equalityPredicates.map(_._1))
-    val predicates = equalityPredicates.map(_._2).toSet
-    def replaceConstants0(expression: Expression) = expression transform {
-      case a: AttributeReference => constantsMap.getOrElse(a, a)
-    }
+  private def replaceConstants(
+      condition: Expression,
+      equalityPredicates: EqualityPredicates): Expression = {
+    val predicates = equalityPredicates.values.map(_._2).toSet
```

Comment:
The reason for changing `EqualityPredicates` to `mutable.Map` earlier was to avoid building a map here.

The main point of that map is that we store only one `Literal` (and its original `BinaryComparision`) assigned to an attribute key. So if we have 2 or more conflicting `EqualTo` then in `replaceConstants()` we keep only one's original form and rewrite the other conflicing ones. E.g. `a = 1 AND a = 2` we store only `a -> (2, a = 2)` in the map and rewrite the expression to `2 = 1 AND a = 2`.